File: backend/v1/app/merge/controller/merge.py
"""音视频合成 API 路由"""
from fastapi import APIRouter, Depends, Path
from sqlalchemy.ext.asyncio import AsyncSession

from backend.store.database.async_database import get_db
from backend.v1.app.merge.dao.merge import (
    AudioReplaceRequest,
    BgmRequest,
    MixRequest,
)
from backend.v1.app.merge.service.merge_service import merge_service
from backend.framework.web import Response
from backend.framework.exceptions import BusinessException
from backend.framework.exceptions.error_codes import RESOURCE_NOT_FOUND, VIDEO_ERROR

router = APIRouter(prefix="/merge", tags=["音视频合成"])


@router.post("/audio-replace", response_model=Response)
async def replace_audio(
    req: AudioReplaceRequest,
    db: AsyncSession = Depends(get_db),
):
    """替换视频音频"""
    try:
        result = await merge_service.replace_audio(
            db, req.video_id, req.audio_id
        )
        return Response.success(data=result)
    except ValueError as e:
        raise BusinessException(VIDEO_ERROR, str(e))


@router.post("/bgm", response_model=Response)
async def add_bgm(
    req: BgmRequest,
    db: AsyncSession = Depends(get_db),
):
    """添加背景音乐"""
    try:
        result = await merge_service.add_bgm(
            db, req.video_id, req.bgm_id,
            req.bgm_volume, req.original_volume
        )
        return Response.success(data=result)
    except ValueError as e:
        raise BusinessException(VIDEO_ERROR, str(e))


@router.post("/mix", response_model=Response)
async def mix_audio(
    req: MixRequest,
    db: AsyncSession = Depends(get_db),
):
    """混合多个音频轨道"""
    try:
        result = await merge_service.mix_audio_tracks(
            db, req.video_id, req.audio_ids, req.volumes
        )
        return Response.success(data=result)
    except ValueError as e:
        raise BusinessException(VIDEO_ERROR, str(e))


@router.get("/tasks/{task_id}", response_model=Response)
async def get_merge_task_status(
    task_id: str = Path(...),
    db: AsyncSession = Depends(get_db),
):
    """查询合成任务状态"""
    try:
        status = await merge_service.get_task_status(db, task_id)
        return Response.success(data=status)
    except ValueError as e:
        raise BusinessException(RESOURCE_NOT_FOUND, str(e))


@router.post("/tasks/{task_id}/cancel", response_model=Response)
async def cancel_merge_task(
    task_id: str = Path(...),
    db: AsyncSession = Depends(get_db),
):
    """取消合成任务"""
    try:
        result = await merge_service.cancel_task(db, task_id)
        return Response.success(data=result)
    except ValueError as e:
        raise BusinessException(VIDEO_ERROR, str(e))

File: frontend/src/services/merge.js
import api from './api'

/**
 * 替换视频音频
 * @param {number} videoId - 视频素材 ID
 * @param {number} audioId - 音频素材 ID
 */
export async function replaceAudio(videoId, audioId) {
  return api.post('/v1/merge/audio-replace', { video_id: videoId, audio_id: audioId })
}

/**
 * 添加背景音乐
 * @param {number} videoId - 视频素材 ID
 * @param {number} bgmId - BGM 素材 ID
 * @param {number} bgmVolume - BGM 音量 (0-1)
 * @param {number} originalVolume - 原始音量 (0-1)
 */
export async function addBgm(videoId, bgmId, bgmVolume = 0.3, originalVolume = 1.0) {
  return api.post('/v1/merge/bgm', {
    video_id: videoId,
    bgm_id: bgmId,
    bgm_volume: bgmVolume,
    original_volume: originalVolume,
  })
}

/**
 * 混合多个音频轨道
 * @param {number} videoId - 视频素材 ID
 * @param {number[]} audioIds - 音频素材 ID 列表
 * @param {number[]} volumes - 各音频音量列表
 */
export async function mixAudio(videoId, audioIds, volumes) {
  return api.post('/v1/merge/mix', {
    video_id: videoId,
    audio_ids: audioIds,
    volumes,
  })
}

/**
 * 查询合成任务状态
 * @param {string} taskId - 任务 ID
 */
export async function getMergeTaskStatus(taskId) {
  return api.get(`/v1/merge/tasks/${taskId}`)
}

/**
 * 取消合成任务
 * @param {string} taskId - 任务 ID
 */
export async function cancelMergeTask(taskId) {
  return api.post(`/v1/merge/tasks/${taskId}/cancel`)
}
